<launch>
    <!-- Pointcloud to Laserscan Node Configuration -->
    <node name="pointcloud_to_laserscan" pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" output="screen">
        <remap from="cloud_in" to="/rgbd"/>
        <remap from="scan" to="/output_scan"/>
        <param name="target_frame" value="frame1"/>
        <param name="min_height" value="-0.5"/>
        <param name="max_height" value="0.5"/>
        <param name="angle_min" value="-3.14"/>
        <param name="angle_max" value="3.14"/>
        <param name="angle_increment" value="0.00314"/>
        <param name="scan_time" value="0.1"/>
        <param name="range_min" value="0"/>
        <param name="range_max" value="10.0"/>
    </node>

    <!-- Slam Toolbox Node Configuration -->
    <node pkg="slam_toolbox" type="sync_slam_toolbox_node" name="slam_toolbox" output="screen">
        <!-- Basic Parameters -->
        <param name="odom_frame" value="odom"/>
        <param name="map_frame" value="world"/>
        <param name="base_frame" value="frame1"/>
        <param name="scan_topic" value="/output_scan"/>
        <param name="mode" value="mapping"/>
        
        <!-- Slam Toolbox Specific Parameters -->
        <param name="solver_plugin" value="solver_plugins::CeresSolver"/>
        <param name="ceres_linear_solver" value="SPARSE_NORMAL_CHOLESKY"/>
        <param name="ceres_preconditioner" value="SCHUR_JACOBI"/>
        <param name="ceres_trust_strategy" value="LEVENBERG_MARQUARDT"/>
        <param name="ceres_dogleg_type" value="TRADITIONAL_DOGLEG"/>
        <param name="ceres_loss_function" value="None"/>
        <param name="resolution" value="0.05"/>
        <param name="max_laser_range" value="10.0"/>
        <param name="minimum_travel_distance" value="0"/>
        <param name="minimum_travel_heading" value="0"/>
        <param name="map_update_interval" value="0.2"/>
        <param name="use_scan_matching" value="false"/>
        <param name="use_scan_barycenter" value="false"/>
        <param name="transform_publish_period" value="0.02"/>
        <param name="debug_logging" value="false"/>
        <param name="throttle_scans" value="1"/>
        <param name="minimum_time_interval" value="0.2"/>
        <param name="transform_timeout" value="0.1"/>
        <param name="tf_buffer_duration" value="30"/>
        <param name="stack_size_to_use" value="40000000"/>
        <param name="use_imu" value="false"/>
        
        <!-- Additional Scan Matcher Parameters -->
        <param name="distance_variance_penalty" value="0.5"/>
        <param name="angle_variance_penalty" value="1.0"/>
        <param name="fine_search_angle_offset" value="0.00349"/>
        <param name="coarse_search_angle_offset" value="0.349"/>
        <param name="coarse_angle_resolution" value="0.0349"/>
        <param name="minimum_angle_penalty" value="0.9"/>
        <param name="minimum_distance_penalty" value="0.5"/>
        <param name="use_response_expansion" value="true"/>
    </node>

    <node name="depth_cloud_to_frame1_transformer" pkg="occupiedmap_rgbd" type="occupymap_rgbd" output="screen">
        <param name="use_outlier_removal" value="true"/>
        <param name="use_pass_through" value="false"/>
        <param name="path_topic" value="/pose_graph/pose_graph_path"/>
        <param name="depth_cloud_topic" value="/pose_graph/orig_cloud2"/>
        <param name="output_cloud_topic" value="/rgbd"/>
        <param name="passthrough_x_min" value="0.2"/>
        <param name="passthrough_x_max" value="20"/>
        <param name="passthrough_z_min" value="-0.5"/>
        <param name="passthrough_z_max" value="1.0"/>
        <param name="sor_mean_k" value="50"/>
        <param name="sor_stddev_thresh" value="1.0"/>

        <param name="window_size" value="21"/>
        <param name="log_enabled" value="true"/>

    </node>

    <node name="pointcloud_converter" pkg="occupiedmap_rgbd" type="pointcloud_to_pointcloud2" output="screen"> </node>
    
    <node name="rviz_map" pkg="rviz" type="rviz" args="-d $(find occupiedmap_rgbd)/rviz_cfg/map.rviz" output="screen"/>
    <node name="rviz_2dscan" pkg="rviz" type="rviz" args="-d $(find occupiedmap_rgbd)/rviz_cfg/2dscan_rgbd.rviz" output="screen"/>
    <node name="rgbd_origin" pkg="rviz" type="rviz" args="-d $(find occupiedmap_rgbd)/rviz_cfg/rgbd_origin.rviz" output="screen"/>
</launch>

